dimod/compatibility23.py

```diff
@@ -35,3 +43,25 @@ def getargspec(f):
         # FullArgSpec(args, varargs, varkw, defaults, kwonlyargs, kwonlydefaults, annotations)
 
         return _ArgSpec(argspec.args, argspec.varargs, argspec.varkw, argspec.defaults)
+
+    # Based on an answer https://stackoverflow.com/a/34757114/8766655
+    # by kindall https://stackoverflow.com/users/416467/kindall
+    class SortKey(object):
```

Comment:
I am confused as to what is the failing case. I do not compare sequences of different types, I only compare them when they are of the same type. 
In python 2:
```
In [1]: [2,1] < (1,1,1)
Out[1]: True

In [2]: (2,1) < (1,1,1)
Out[2]: False

In [3]: [2,1] < [1,1,1]
Out[3]: False

In [4]: [2,1] < (1,1,1)
Out[4]: True
```
in python3:
```
In [4]: SortKey([2,1])<SortKey((1,1,1))
Out[4]: True

In [5]: SortKey((2,1)) < SortKey((1,1,1))
Out[5]: False

In [6]: (2,1) < (1,1,1)
Out[6]: False

In [7]: SortKey([2,1]) < SortKey([1,1,1])
Out[7]: False

In [8]: [2,1] < [1,1,1]
Out[8]: False

In [9]: SortKey([2,1]) < SortKey((1,1,1))
Out[9]: True
```

Reply:
Oh, I overlooked the part about comparing same type sequences (because that doesn't make sense - if both are sequences, then why insist on the same type? Non-generator sequences are already ordered by default, e.g. lists and tuples).

But the code is still incorrect for the case when sequences are actually compared (for loop).

Consider in Python2:
```
>>> range(2,4) < range(0,3)
False
```
and in Python3:
```
>>> SortKey(range(2,4)) < SortKey(range(0,3))                                                      
True
```

Surely it would make sense for `[2,3]` not to be less than `[0,1,2]`. And actually:
```
>>> SortKey(list(range(2,4))) < SortKey(list(range(0,3)))                                          
False
```

Your **code fails because in case the first sequence (element by element) is not less then the second, you failback to comparing lengths** (which you should only if one sequence is exact prefix of the other).

Reply:
We check that they are the same type because otherwise we can sort them by type name (like in python2). This additional code is meant to check the case that it has nested objects that need to be compared using sortkeys, otherwise it would pass the original try-catch. Agree the logic needs to change for the range case.